### PR TITLE
[Fix] fix some neo issues

### PR DESCRIPTION
**Pull request by w20089527**
## Description
The changes of this pr are on bug fix and some code refactoring.
### Bug fix (non-breaking changes)
1. fix neo txid & asset id serialization missing zero leading issue
2. fix neo output check in plan not correct
3. fix TWCardanoGetStakingAddress not included issue
4. fix neo attribute serialize & deserialize not correct
   - TAU_DescriptionUrl
   - TAU_Description
   - TAU_Remark

### Code refactoring
1. use `switch` instead of `if-else` in `TransactionAttribute.h`
2. use constant variables instead of number directly

## Types of changes

Bug fix (non-breaking changes)
- [fix neo txid & asset id serialization missing zero leading issue](https://github.com/trustwallet/wallet-core/pull/2599/commits/cc8ff67e01b03e3067bcf3d19280a19e241946f8)
- [fix neo output check in plan not correct](https://github.com/trustwallet/wallet-core/pull/2599/commits/7ac1552aced800fc1134a48928ca5a7aa1a4c810)
- [fix TWCardanoGetStakingAddress not included issue](https://github.com/trustwallet/wallet-core/pull/2599/commits/a8badde56ef45e6ea2385d6c8d522447abfe5b76)
- [fix neo attribute serialize & deserialize not correct](https://github.com/trustwallet/wallet-core/pull/2599/commits/e2d9c36c476ebf4fcfe0a0c1e76aa3e303a6cef2)
  - TAU_DescriptionUrl
  - TAU_Description
  - TAU_Remark

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: src/NEO/CoinReference.h [hunk 1]
```diff
@@ -7,7 +7,7 @@
 #pragma once
 
 #include "../uint256.h"
-#include "Data.h"
```
Comment (hewigovens) [inline src/NEO/CoinReference.h:10]: hmm, This should just work

File: src/NEO/TransactionAttribute.h [hunk 2]
```diff
@@ -1,54 +1,85 @@
...
-  public:
+public:
     TransactionAttributeUsage usage = TAU_ContractHash;
     Data _data;
 
     virtual ~TransactionAttribute() {}
 
     int64_t size() const override {
-        return 1 + _data.size();
+        if (usage == TransactionAttributeUsage::TAU_ContractHash ||
```
Comment (hewigovens) [inline src/NEO/TransactionAttribute.h:24]: use `switch` instead

File: tests/NEO/TransactionAttributeTests.cpp [hunk 3]
```diff
@@ -84,8 +86,9 @@ TEST(NEOTransactionAttribute, Deserialize) {
 TEST(NEOTransactionAttribute, DeserializeInitialPositionAfterData) {
     auto transactionAttribute = TransactionAttribute();
     EXPECT_THROW(transactionAttribute.deserialize(Data(), 1), std::invalid_argument);
-
+
```
Comment (hewigovens) [inline tests/NEO/TransactionAttributeTests.cpp:89]:
```suggestion
```

**Review by hewigovens — COMMENTED** [review]
left some comments:
1. "Data.h" should work
2. use `switch` for `TransactionAttributeUsage`

File: src/NEO/CoinReference.h (hunk 1 above)
Comment (w20089527) [reply src/NEO/CoinReference.h:10]: Yes, just want to keep same style with others.

Comment (catenocrypt) [reply src/NEO/CoinReference.h:10]: Funny, if you look at the last change to this file, "../Data.h" was changed to "Data.h", but across the whole project.
https://github.com/trustwallet/wallet-core/pull/2563
I suggest to change to without the ".." format the other includes as well.

Comment (w20089527) [reply src/NEO/CoinReference.h:10]: Awesome, I will fix it.

**Review by hewigovens — APPROVED** [review]

**Review by Milerius — DISMISSED** [review]

**Review by miloserdow — APPROVED** [review]

File: src/NEO/CoinReference.h [hunk 4]
```diff
@@ -35,7 +35,7 @@ class CoinReference : public Serializable {
     }
 
     Data serialize() const override {
-        auto resp = store(prevHash);
+        auto resp = store(prevHash, 32);
```
Comment (catenocrypt) [inline src/NEO/CoinReference.h:38]: Could you please use a constant for 32 here?

File: src/NEO/TransactionAttribute.h [hunk 5]
```diff
@@ -1,54 +1,100 @@
...
     virtual ~TransactionAttribute() {}
 
     int64_t size() const override {
-        return 1 + _data.size();
+        switch (usage) {
+        case TransactionAttributeUsage::TAU_ContractHash:
+        case TransactionAttributeUsage::TAU_ECDH02:
+        case TransactionAttributeUsage::TAU_ECDH03:
+        case TransactionAttributeUsage::TAU_Vote:
+            return 1 + 32;
```
Comment (catenocrypt) [inline src/NEO/TransactionAttribute.h:29]: Constant here? Maybe assetIdSize

File: src/NEO/TransactionAttribute.h [hunk 6]
```diff
@@ -1,54 +1,100 @@
...
+        case TransactionAttributeUsage::TAU_Script:
+            return 1 + 20;
```
Comment (catenocrypt) [inline src/NEO/TransactionAttribute.h:31]: Constant here? Maybe scriptHashSize

File: src/NEO/TransactionAttribute.h [hunk 7]
```diff
@@ -1,54 +1,100 @@
...
+        default:
+            if (usage >= TransactionAttributeUsage::TAU_Hash1 &&
+                usage <= TransactionAttributeUsage::TAU_Hash15) {
+                return 1 + 32;
```
Comment (catenocrypt) [inline src/NEO/TransactionAttribute.h:35]: Same here

File: src/NEO/TransactionAttribute.h [hunk 8]
```diff
@@ -1,54 +1,100 @@
...
+
+        // see: https://github.com/neo-project/neo/blob/v2.12.0/neo/Network/P2P/Payloads/TransactionAttribute.cs#L49
+        if (usage == TransactionAttributeUsage::TAU_DescriptionUrl ||
+            usage == TransactionAttributeUsage::TAU_Description ||
+            usage >= TransactionAttributeUsage::TAU_Remark) {
+            Data resp;
+            encodeVarInt((uint64_t)_data.size(), resp);
+            result.insert(result.end(), resp.begin(), resp.end());
+        }
+        if (usage == TransactionAttributeUsage::TAU_ECDH02 ||
```
Comment (catenocrypt) [inline src/NEO/TransactionAttribute.h:90]: 'else if' here if it's correct?

**Review by catenocrypt — CHANGES_REQUESTED** [review]

File: src/NEO/TransactionAttribute.h (hunk 8 above)
Comment (w20089527) [reply src/NEO/TransactionAttribute.h:90]: That's expected.

File: src/NEO/Constants.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+#pragma once
+
+namespace TW::NEO {
+
+static const size_t assetIdSize = 32;
+static const size_t contractHashSize = 32;
+static const size_t valueSize = 8;
+static const size_t scriptHashSize = 20;
+
+} // namespace TW::NEO
```
Comment (hewigovens) [inline src/NEO/Constants.h:16]: can you add an empty line to the end of each file changed?

Comment (w20089527) [reply src/NEO/Constants.h]: done

**Review by hewigovens — APPROVED** [review]

**Review by catenocrypt — APPROVED** [review]

**Pull request merged**